design-proposals/nic-hotplug/nic-hotplug.md

```diff
@@ -0,0 +1,407 @@
+An important detail to take into account is this controller **must** be
+executed on the host's PID, network, and mount namespaces, in order to access
+whatever resources the delegate CNI requires which are available in the host
+file system (e.g. log files, sockets, etc). The alternative would be for the
+multus configuration to somehow know which resources to bind mount into the pod,
+which, in my opinion is not reasonable without major changes in multus - we'd
+need some discovery mechanism between multus & the CNI plugins.
+
+There are multiple alternatives to run the controller process:
+1. the controller simply listens for pod network annotation updates, and spawns
+   a new process (the "thin" CNI plugin) that runs on the host's PID, net, and
+   mount namespaces.
+2. multus is re-architected as a thick plugin, which would require all its code
+   to run in host's mount namespaces. This solution implies a larger attack
+   surface on multus.
+
+Both these alternatives **require** multus to be granted the ability of sharing
+the PID namespace of the host where they run.
+
+### Simple controller
+This controller would just start a new process on the host's mount namespace.
+
+To achieve it, the following code snippet could be explorer:
+
+```golang
+hostMountNamespace := "/proc/1/ns/mnt"
+fd, err := os.Open(hostMountNamespace)
+if err != nil {
+    return fmt.Errorf("failed to open mount namespace: %v", err)
+}
+defer fd.Close()
+
+if err = unix.Unshare(unix.CLONE_NEWNS); err != nil {
+    return fmt.Errorf("failed to detach from parent mount namespace: %v", err)
+}
+if err := unix.Setns(int(fd.Fd()), unix.CLONE_NEWNS); err != nil {
+    return fmt.Errorf("failed to join the mount namespace: %v", err)
+}
+
+pathToCNIBinary := "/opt/cni/bin/" + cniPluginName
+cniCmd := exec.Command(pathToCNIBinary)
+cniCmd.Env = <the environment we want to pass>
+result, err := cniCmd.Run()
+...
+```
+
+### Thickening multus
+A "thin CNI plugin" runs as a one-shot process, typically as a binary on disk
+executed on a Kubernetes host machine.
+
+A "thick CNI Plugin", on the other hand, is a CNI component composed of two (or
+more) parts, usually composed of "shim", and a long lived process (daemon)
+resident in memory. The "shim" is a lightweight "thin CNI plugin" component that
+simply passes CNI parameters (such as JSON configuration, and environment
+variables) to the daemon component, which then processes the CNI request.
+
+To transform multus into a thick plugin, it is needed to instantiate a long
+lived process - which will be the multus pod entrypoint - listening to a unix
+domain socket - this socket must be available both in the multus pod and the
+hosts's mount namespaces; as such, a bind mount to host this socket must be
+provided for the multus pod.
+
+The CNI shim will then be invoked by kubelet, and send the CNI ADD/DELETE
+commands to the server side of multus via the unix domain socket previously
+mentioned.
+
```

Comment:
Is there any concerns for CMD_DEL? In container/Pod case, what is happen if the net-attach-def, which contains default route, is removed? How to get new default route?

Reply:
Looking at kaloom's code, I see that they send `CNI_DEL` for each pod connected to the NAD that got removed: https://github.com/kaloom/kubernetes-podagent#deleting-a-network-attachment . IMO, that would orphan out the running pod default route. I guess it might not be that big of a deal, since - iiuc - you have to be an admin to remove NADs. 

I do not know what is more correct; to do that, **or** have a validating webhook that prevents the deletion of a network **whenever** there are pods attached to it. I feel inclined to say the latter is the preferable choice, but I'm not sure. 

I think the community has a say in this. We could take this to this week's community call.

Reply:
multus has 'default-route' feature in network annotation, which allows user to change default route to net-attach-def network. As far as I know of, Kactus does not have this feature.

If user pod connects to 'net-attach1', 'net-attach2' and 'net-attach3' with 'cluster network' and pod's default route is to 'net-attach2', then if user changes pod network to remove 'net-attach2'. In this case, user want not only to remove default route, also to change default route to cluster network. It does not make it by just invoke `CMD_DEL`.

Could you please take into account it?

Reply:
Oh I see, I had misunderstood your comment. 

What if the user want the default route in another network ? Reacting to NAD updates is not in scope of this feature for KubeVirt. IMO, the scenario you point out is a corner case of that. 

I think the fairest option here is to either document the limitation, or create a follow-up story in multus to address this scenario. 

Out of curiosity, how would this be done technically ? Re-run the primary/default CNI plugin indicating that it is now the default route ?